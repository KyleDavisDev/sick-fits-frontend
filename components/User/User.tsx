import * as React from "react";
import { Query } from "react-apollo";
import gql from "graphql-tag";
import { ReactElement } from "react";

export const CURRENT_USER_QUERY = gql`
  query CURRENT_USER_QUERY {
    me {
      id
      email
      name
      permissions
      cart {
        created
        id
        items {
          id
          quantity
          item {
            id
            price
            image
            title
            description
          }
        }
      }
    }
  }
`;

export interface IUserProps {
  children(data: any): ReactElement;
}

const User: React.FunctionComponent<IUserProps> = props => {
  return (
    <Query {...props} query={CURRENT_USER_QUERY}>
      {payload => {
        return props.children(payload);
      }}
    </Query>
  );
};

export default User;
