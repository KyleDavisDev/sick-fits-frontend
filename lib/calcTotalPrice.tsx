export default function(items) {
  return items.reduce((tally, cartItem) => {
    if (!cartItem.item) return tally;
    return tally + cartItem.quantity * cartItem.item.price;
  }, 0);
}
