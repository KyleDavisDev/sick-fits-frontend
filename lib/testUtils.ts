import casual from "casual";
import { ReactWrapper } from "enzyme";

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  __typename: "Item",
  created: casual.unix_time,
  id: casual.uuid,
  price: 5000,
  user: null,
  image: "dog-small.jpg",
  title: "dogs are best",
  description: "dogs",
  largeImage: "dog.jpg"
});

const fakeUser = () => ({
  __typename: "User",
  id: casual.uuid,
  name: casual.name,
  email: casual.email,
  permissions: ["ADMIN"],
  orders: [],
  cart: []
});

const fakeOrderItem = () => ({
  __typename: "OrderItem",
  id: casual.uuid,
  image: `${casual.word}.jpg`,
  title: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words()
});

const fakeOrder = () => ({
  __typename: "Order",
  id: casual.uuid,
  charge: "ch_123",
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: "2018-04 - 06T19: 24: 16.000Z",
  user: fakeUser()
});

const fakeCartItem = (overrides?: any) => ({
  __typename: "CartItem",
  id: casual.uuid,
  created: casual.unix_time,
  quantity: casual.integer(0, 15),
  item: fakeItem(),
  user: fakeUser(),
  ...overrides
});

const fakeCart = (overrides?: any) => ({
  __typename: "Cart",
  id: casual.uuid,
  items: [fakeCartItem(), fakeCartItem(), fakeCartItem()],
  user: fakeUser(),
  created: casual.unix_time,
  updated: casual.unix_time,
  isActive: true,
  ...overrides
});

// Fake LocalStorage
class LocalStorageMock {
  store: any;
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

function simulateInputChange(
  wrapper: ReactWrapper,
  name: string,
  value: string
) {
  wrapper
    .find(`input[name='${name}']`)
    .simulate("change", { target: { name, value } });
}

export {
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeCart,
  fakeOrder,
  fakeOrderItem,
  simulateInputChange
};
