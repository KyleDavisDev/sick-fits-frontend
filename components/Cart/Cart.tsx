import * as React from "react";
import { Query, Mutation } from "react-apollo";
import gql from "graphql-tag";
import { StyledDiv, StyledHeader, StyledCloseButton } from "./CartStyles";
import Button from "../Button/Button";
import User from "../User/User";
import CartItem from "../CartItem/CartItem";

export const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`;

export const TOGGLE_CART_MUTATION = gql`
  mutation {
    toggleCart @client
  }
`;

export interface ICartProps {}

const Cart: React.FunctionComponent<ICartProps> = (props: ICartProps) => {
  return (
    <User>
      {({ data: { me } }) => {
        if (!me) return null;
        return (
          <Mutation mutation={TOGGLE_CART_MUTATION}>
            {toggleCart => {
              return (
                <Query query={LOCAL_STATE_QUERY}>
                  {({ data }) => {
                    return (
                      <StyledDiv open={data.cartOpen}>
                        <header>
                          <StyledCloseButton
                            title={"close"}
                            onClick={toggleCart}
                          >
                            &times;
                          </StyledCloseButton>
                          <StyledHeader>{me.name} Cart</StyledHeader>
                          <p>
                            You have{" "}
                            {me.cart && me.cart.items
                              ? me.cart.items.length
                              : 0}{" "}
                            item
                            {me.cart &&
                            me.cart.items &&
                            me.cart.items.length > 1
                              ? "s"
                              : ""}{" "}
                            in your cart.
                          </p>
                        </header>
                        <ul>
                          {me.cart.items.map(cartItem => {
                            return (
                              <CartItem key={cartItem.id}>
                                {cartItem.id}
                              </CartItem>
                            );
                          })}
                        </ul>
                        <footer>
                          <p>$10.00</p>
                          <Button onClick={() => {}}>Checkout</Button>
                        </footer>
                      </StyledDiv>
                    );
                  }}
                </Query>
              );
            }}
          </Mutation>
        );
      }}
    </User>
  );
};

export default Cart;
