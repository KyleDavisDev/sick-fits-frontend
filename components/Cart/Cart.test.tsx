import "jsdom-global/register";
import * as React from "react";
import { mount } from "enzyme";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";
import { ApolloConsumer } from "react-apollo";
import Cart, { LOCAL_STATE_QUERY } from "./Cart";
import { CURRENT_USER_QUERY } from "../User/User";
import { fakeUser, simulateInputChange, fakeCart } from "../../lib/testUtils";

const me = fakeUser();
const mocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me } }
  },
  {
    request: { query: LOCAL_STATE_QUERY },
    result: { data: { cartOpen: true } }
  }
];

describe("<Cart />", () => {
  it("renders and matches snapshot", async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Cart />
      </MockedProvider>
    );

    await wait();
    wrapper.update();

    // check snaps
    expect(wrapper.find("header p")).toMatchSnapshot();
    expect(wrapper.find("header h3")).toMatchSnapshot();
  });
});
