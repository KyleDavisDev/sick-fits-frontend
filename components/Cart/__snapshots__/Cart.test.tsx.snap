// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> renders and matches snapshot 1`] = `
<p>
  You have 
  3
   
  item
  s
   
  in your cart.
</p>
`;

exports[`<Cart /> renders and matches snapshot 2`] = `
<h3
  className="sc-bwzfXH eqnxQe"
>
  Miss Katrine Schultz
   Cart
</h3>
`;

exports[`<Cart /> renders and matches snapshot 3`] = `
<p>
  $900
</p>
`;
